File: src/main/kotlin/fr/jhelp/kotlinLigthTools/promise/Future.kt
package fr.jhelp.kotlinLigthTools.promise

import fr.jhelp.kotlinLight.CommonList
import fr.jhelp.kotlinLight.Mutex
import fr.jhelp.kotlinLight.Try
import fr.jhelp.kotlinLigthTools.atomic.AtomicReference
import fr.jhelp.kotlinLigthTools.tasks.launchTask

class Future<T>
{
    private val promise: Promise<T>
    private val futureStatus: AtomicReference<FutureStatus>
    private val mutex: Mutex
    private var result: T?
    private var error: String
    private var continuations: CommonList<() -> Unit>

    internal constructor(promise: Promise<T>)
    {
        this.promise = promise
        this.futureStatus = AtomicReference<FutureStatus>(FutureStatus.COMPUTING)
        this.mutex = Mutex()
        this.result = null
        this.error = ""
        this.continuations = CommonList<() -> Unit>()
    }

    fun status(): FutureStatus
    {
        return this.futureStatus.get()!!
    }

    fun getResult(): T?
    {
        return this.result
    }

    fun getError(): String
    {
        return this.error
    }

    fun getCancelReason(): String?
    {
        return this.promise.getCancelReason()
    }

    fun cancel(reason: String)
    {
        if (this.futureStatus.get() == FutureStatus.COMPUTING)
        {
            this.futureStatus.set(FutureStatus.CANCELED)
            this.promise.cancel(reason)
            this.fireListeners()
        }
    }

    fun <T1> andThen(continuation: (T) -> T1): Future<T1>
    {
        val promise = Promise<T1>()
        val action: () -> Unit = {
            when (this.futureStatus.get()!!)
            {
                FutureStatus.SUCCEED ->
                {
                    try
                    {
                        @Try promise.result(continuation(this.result!!))
                    }
                    catch (error: Exception)
                    {
                        promise.error("Failed to execute continuation : $error")
                    }
                }
                FutureStatus.FAILED -> promise.error(this.error)
                FutureStatus.CANCELED -> promise.getFuture().cancel(this.getCancelReason()!!)
                else                  ->
                {
                }

            }
        }

        this.mutex.safeExecute {
            when (this.futureStatus.get()!!)
            {
                FutureStatus.COMPUTING -> this.continuations.append(action)
                else                   -> launchTask(action)
            }
        }

        return promise.getFuture()
    }

    fun <T1> then(continuation: (Future<T>) -> T1): Future<T1>
    {
        val promise = Promise<T1>()
        val action: () -> Unit = {
            try
            {
                @Try promise.result(continuation(this))
            }
            catch (error: Exception)
            {
                promise.error("Failed to execute continuation : $error")
            }
        }

        this.mutex.safeExecute {
            when (this.futureStatus.get()!!)
            {
                FutureStatus.COMPUTING -> this.continuations.append(action)
                else                   -> launchTask(action)
            }
        }

        return promise.getFuture()
    }

    fun <T1> andThenUnwrap(continuation: (T) -> Future<T1>): Future<T1>
    {
        return unwrap(this.andThen(continuation))
    }

    fun <T1> thenUnwrap(continuation: (Future<T>) -> Future<T1>): Future<T1>
    {
        return unwrap(this.then(continuation))
    }

    internal fun result(result: T)
    {
        if (this.futureStatus.get() == FutureStatus.COMPUTING)
        {
            this.futureStatus.set(FutureStatus.SUCCEED)
            this.result = result
            this.fireListeners()
        }
    }

    internal fun error(cause: String)
    {
        if (this.futureStatus.get() == FutureStatus.COMPUTING)
        {
            this.futureStatus.set(FutureStatus.FAILED)
            this.error = cause
            this.fireListeners()
        }
    }

    private fun fireListeners()
    {
        this.mutex.safeExecute {
            for (continuation in this.continuations)
            {
                launchTask(continuation)
            }

            this.continuations.removeAll()
        }
    }
}

File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/promise/Future.swift


public class Future<T> {

    private let promise: Promise<T>
    private let futureStatus: AtomicReference<FutureStatus>
    private let mutex: Mutex
    private var result: T?
    private var error: String
    private var continuations: Array<() -> Void>

    internal init(_ promise: Promise<T>){
        self.promise = promise
        self.futureStatus = AtomicReference<FutureStatus>(FutureStatus.COMPUTING)
        self.mutex = Mutex()
        self.result = nil
        self.error = ""
        self.continuations = Array<() -> Void>()
    }

    public func status() -> FutureStatus
    {
        return self.futureStatus.get()!
    }

    public func getResult() -> T?
    {
        return self.result
    }

    public func getError() -> String
    {
        return self.error
    }

    public func getCancelReason() -> String?
    {
        return self.promise.getCancelReason()
    }

    public func cancel(_ reason: String)
    {
        if self.futureStatus.get() == FutureStatus.COMPUTING
        {
            self.futureStatus.set(FutureStatus.CANCELED)
            self.promise.cancel(reason)
            self.fireListeners()
        }
    }

    public func andThen <T1>(_ continuation: @escaping  (T) -> T1) -> Future<T1>
    {
        let promise = Promise<T1>()
        let action: () -> Void = {
            switch self.futureStatus.get()!
            {
                case FutureStatus.SUCCEED  : 
                do {
                    do
                    {
                        try promise.result(continuation(self.result!))
                    }
                    catch
                    {
                        promise.error("Failed to execute continuation : \(error)")
                    }
                }
                case FutureStatus.FAILED  :  promise.error(self.error)
                case FutureStatus.CANCELED  :  promise.getFuture().cancel(self.getCancelReason()!)
                default                   : 
                do {
                }

            }
        }

        self.mutex.safeExecute {
            switch self.futureStatus.get()!
            {
                case FutureStatus.COMPUTING  :  self.continuations.append(action)
                default                    :  launchTask(action)
            }
        }

        return promise.getFuture()
    }

    public func then <T1>(_ continuation: @escaping  (Future<T>) -> T1) -> Future<T1>
    {
        let promise = Promise<T1>()
        let action: () -> Void = {
            do
            {
                try promise.result(continuation(self))
            }
            catch
            {
                promise.error("Failed to execute continuation : \(error)")
            }
        }

        self.mutex.safeExecute {
            switch self.futureStatus.get()!
            {
                case FutureStatus.COMPUTING  :  self.continuations.append(action)
                default                    :  launchTask(action)
            }
        }

        return promise.getFuture()
    }

    public func andThenUnwrap <T1>(_ continuation: @escaping  (T) -> Future<T1>) -> Future<T1>
    {
        return unwrap(self.andThen(continuation))
    }

    public func thenUnwrap <T1>(_ continuation: @escaping  (Future<T>) -> Future<T1>) -> Future<T1>
    {
        return unwrap(self.then(continuation))
    }

    internal func result(_ result: T)
    {
        if self.futureStatus.get() == FutureStatus.COMPUTING
        {
            self.futureStatus.set(FutureStatus.SUCCEED)
            self.result = result
            self.fireListeners()
        }
    }

    internal func error(_ cause: String)
    {
        if self.futureStatus.get() == FutureStatus.COMPUTING
        {
            self.futureStatus.set(FutureStatus.FAILED)
            self.error = cause
            self.fireListeners()
        }
    }

    private func fireListeners()
    {
        self.mutex.safeExecute {
            for continuation in self.continuations
            {
                launchTask(continuation)
            }

            self.continuations.removeAll()
        }
    }
}
